components/external_storage/src/locking.rs

```diff
@@ -0,0 +1,416 @@
+//! encounter live locks when there are too many clients racing for the same
+//! lock.
+
+use std::io;
+
+use chrono::Utc;
+use futures_util::{
+    future::{ok, FutureExt, LocalBoxFuture},
+    io::AsyncReadExt,
+    stream::TryStreamExt,
+};
+use tikv_util::sys::{
+    hostname,
+    thread::{process_id, Pid},
+};
+use uuid::Uuid;
+
+use crate::{ExternalStorage, UnpinReader};
+
+#[derive(serde::Serialize, serde::Deserialize)]
+struct LockMeta {
+    locked_at: chrono::DateTime<Utc>,
+    locker_host: String,
+    locker_pid: Pid,
+    txn_id: Uuid,
+    hint: String,
+}
+
+impl LockMeta {
+    fn new(txn_id: Uuid, hint: String) -> Self {
+        Self {
+            locked_at: Utc::now(),
+            locker_host: hostname().unwrap_or_else(|| "an_unknown_tikv_node".to_owned()),
+            locker_pid: process_id(),
+            txn_id,
+            hint,
+        }
+    }
+
+    fn to_json(&self) -> io::Result<Vec<u8>> {
+        serde_json::ser::to_vec(self).map_err(|err| io::Error::new(io::ErrorKind::InvalidData, err))
+    }
+}
+
+#[derive(Debug)]
+pub struct RemoteLock {
+    txn_id: uuid::Uuid,
+    path: String,
+}
+
+impl RemoteLock {
+    /// Unlock this lock.
+    ///
+    /// If the lock was modified, this may return an error.
+    pub async fn unlock(&self, st: &dyn ExternalStorage) -> io::Result<()> {
+        let mut buf = vec![];
+        st.read(&self.path).read_to_end(&mut buf).await?;
+        let meta = serde_json::from_slice::<LockMeta>(&buf)?;
+        if meta.txn_id != self.txn_id {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                format!(
+                    "TXN ID mismatch, remote {} ours {}",
+                    meta.txn_id, self.txn_id
+                ),
+            ));
+        }
+
+        st.delete(&self.path).await
+    }
+}
+
+/// The [`ExclusiveWriteTxn`] instance for putting a read lock for a path.
+struct PutRLock {
+    basic_path: String,
+    lock_path: String,
+    hint: String,
+}
+
+impl PutRLock {
+    fn new(path: &str, hint: String) -> Self {
+        Self {
+            basic_path: path.to_string(),
+            lock_path: format!("{}.READ.{:016x}", path, rand::random::<u64>()),
+            hint,
+        }
+    }
+}
+
+impl ExclusiveWriteTxn for PutRLock {
+    fn path(&self) -> &str {
+        &self.lock_path
+    }
+
+    fn content(&self, cx: ExclusiveWriteCtx<'_>) -> io::Result<Vec<u8>> {
+        LockMeta::new(cx.txn_id(), self.hint.clone()).to_json()
+    }
+
+    fn verify<'cx: 'ret, 's: 'ret, 'ret>(
+        &'s self,
+        cx: ExclusiveWriteCtx<'cx>,
+    ) -> LocalBoxFuture<'ret, io::Result<()>> {
+        // We need capture `cx` here, or rustc complains that we are returning a future
+        // reference to a local variable. (Yes indeed.)
+        async move {
+            cx.check_files_of_prefix(&format!("{}.WRIT", self.basic_path), requirements::nothing)
```

Comment:
What are files  in the `{}.WRIT`?

Reply:
OK, I get it. I wonder whether the number of files in the `basic_path` is large so that lock file iterate operation will be heavy.

Reply:
I think this should fail fast when it encounters the first file contains the prefix.